virtwho/virt/vdsm/vdsm.py

```diff
@@ -107,46 +112,120 @@ def _get_local_vds_name(self, trusted_store_path):
+            return True
+        except TypeError:
+            return False
 
+    def _create_ssl_context(self):
         key_file = self.trust_store_path + '/keys/vdsmkey.pem'
         cert_file = self.trust_store_path + '/certs/vdsmcert.pem'
         ca_cert = self.trust_store_path + '/certs/cacert.pem'
 
-        ctx = SSL.Context()
+        if self._need_m2crypto():
+            ctx = ssl.create_default_context(capath=ca_cert)
+            ctx.load_cert_chain(cert_file, key_file)
+            ctx.check_hostname = False
+            ctx.verify_mode = ssl.CERT_NONE
+        else:
+            from M2Crypto import SSL
 
-        ctx.set_verify(SSL.verify_peer | SSL.verify_fail_if_no_peer_cert, 16)
-        ctx.load_verify_locations(ca_cert)
-        ctx.load_cert(cert_file, key_file)
+            ctx = SSL.Context()
 
-        return xmlrpclib.Server('https://%s:%s' % (addr, self.management_port), SSL_Transport(ctx))
+            ctx.set_verify(SSL.verify_peer | SSL.verify_fail_if_no_peer_cert, 16)
+            ctx.load_verify_locations(ca_cert)
+            ctx.load_cert(cert_file, key_file)
+        return ctx
 
     def connect(self):
         if self.trust_store_path:
-            try:
-                self.server = self._secure_connect()
+            ssl_context = self._create_ssl_context()
+            addr = self._get_local_vds_name(self.trust_store_path)
+            client = self._jsonrpc(addr, ssl_context) or self._xmlrpc(addr, ssl_context)
+            if client:
                 return
-            except Exception as e:
-                self.logger.exception(e)
-        # Try http version if ssl is off or fails
-        self.server = xmlrpclib.Server("http://localhost:%s" % self.management_port)
+
+        # Try http versions if ssl is off or fails
+        self._jsonrpc('localhost') or self._xmlrpc('localhost')
+
+    def _jsonrpc(self, addr, ssl_context=None):
+        try:
+            jsonrpc_client = JsonRpcClient(addr, self.management_port, ssl_context, timeout=VDSM_JSONRPC_TIMEOUT)
+            jsonrpc_client.connect()
+            self.jsonrpc_client = jsonrpc_client
+            return True
+        except IOError:
+            if ssl_context:
+                description = 'JSON-RPC with SSL'
+            else:
+                description = 'JSON-RPC'
+            self.logger.warning('Unable to connect via %s' % description, exc_info=True)
+        return False
+
+    def _xmlrpc(self, addr, ssl_context=None):
+        try:
+            transport = None
+            if ssl_context:
+                if self._need_m2crypto():
+                    transport = xmlrpclib.SafeTransport(context=ssl_context)
+                else:
+                    from M2Crypto.m2xmlrpclib import SSL_Transport
+                    transport = SSL_Transport(ssl_context)
+                uri = 'https://%s:%s' % (addr, self.management_port)
+            else:
+                uri = 'http://%s:%s' % (addr, self.management_port)
```

Comment:
It looks like that we are enabling unsecured connection to remote server here. On the other side it seems that `_xmlrpc()` is called only with `addr='localhost'`, when `ssl_context=None`.